### PR TITLE
Add move_vm_from_on_prem_to_aws playbook and corresponding roles

### PR DESCRIPTION
**Pull request by alinabuzachis**
Add move_vm_from_on_prem_to_aws playbook and corresponding roles

Requires https://github.com/ansible-collections/amazon.aws/pull/1723

### CODE REVIEW

File: playbooks/move_vm_from_on_prem_to_aws/vars/main.yml
Comment (gravesm) [inline playbooks/move_vm_from_on_prem_to_aws/vars/main.yml:1]: Was this file intended to be checked in? These values seem very specific to a local testing environment.

File: playbooks/move_vm_from_on_prem_to_aws/move_vm_from_on_prem_to_aws.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        clone_on_prem_vm_source_vm_name: "{{ on_prem_source_vm_name }}"
+        clone_on_prem_vm_dest_vm_name: "{{ on_prem_vm_clone_name }}"
+        clone_on_prem_vm_uri: "{{ uri }}"
+        clone_on_prem_vm_local_image_path: "{{ local_image_path }}"
+      delegate_to: kvm
+
+    - name: Import 'cloud.aws_ops.import_image_and_run_aws_instance' role
+      ansible.builtin.import_role:
+        name: cloud.aws_ops.import_image_and_run_aws_instance
+      vars:
```
Comment (gravesm) [inline playbooks/move_vm_from_on_prem_to_aws/move_vm_from_on_prem_to_aws.yml:37]: It looks like there are some variables missing here. How does the user set the security groups, keypair, subnet, etc.?

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+* **instance_type** (str): (Optional) The EC2 instance type you want to use. Default: "t2.micro".
+* **keypair_name** (str): (Optional) The name of the SSH access key to assign to the EC2 instance. It must exist in the region the instance is created. If not set, your default AWS account keypair will be used.
+* **security_groups** (list): (Optional) A list of security group IDs or names to assiciate to the EC2 instance.
+* **vpc_subnet_id** (str): (Optional) The subnet ID in which to launch the EC2 instance instance (VPC). If none is provided, M(amazon.aws.ec2_instance) will chose the default zone of the default VPC.
+* **instance_volumes** (dict): (Optional) A dictionary of a block device mappings, by default this will always use the AMI root device so the **instance_volumes** option is primarily for adding more storage. A mapping contains the (optional) keys _device_name_, _ebs.volume_type_, _ebs.volume_size_, _ebs.kms_key_id_, _ebs.iops_, and _ebs.delete_on_termination_.
+
+* **kvm_host** (str): Information about the host running the KVM hypervisr that are dynamically added to the inventory.
```
Comment (gravesm) [inline playbooks/move_vm_from_on_prem_to_aws/README.md:57]:
```suggestion
* **kvm_host** (dict): Information about the host running the KVM hypervisr that are dynamically added to the inventory.
```

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+* **image_path** (str): (Required) The path where the .raw image is stored.
+* **instance_name** (str): (Required) The name of the EC2 instance you want to create using the imported AMI.
+* **instance_type** (str): (Optional) The EC2 instance type you want to use. Default: "t2.micro".
+* **keypair_name** (str): (Optional) The name of the SSH access key to assign to the EC2 instance. It must exist in the region the instance is created. If not set, your default AWS account keypair will be used.
+* **security_groups** (list): (Optional) A list of security group IDs or names to assiciate to the EC2 instance.
+* **vpc_subnet_id** (str): (Optional) The subnet ID in which to launch the EC2 instance instance (VPC). If none is provided, M(amazon.aws.ec2_instance) will chose the default zone of the default VPC.
+* **instance_volumes** (dict): (Optional) A dictionary of a block device mappings, by default this will always use the AMI root device so the **instance_volumes** option is primarily for adding more storage. A mapping contains the (optional) keys _device_name_, _ebs.volume_type_, _ebs.volume_size_, _ebs.kms_key_id_, _ebs.iops_, and _ebs.delete_on_termination_.
```
Comment (gravesm) [inline playbooks/move_vm_from_on_prem_to_aws/README.md:55]: This would benefit from having the same structure as you use for the `kvm_host` var.

File: playbooks/move_vm_from_on_prem_to_aws/vars/main.yml
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/vars/main.yml:1]: Yes, it was intended to show some examples.

File: playbooks/move_vm_from_on_prem_to_aws/move_vm_from_on_prem_to_aws.yml (hunk 1 above)
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/move_vm_from_on_prem_to_aws.yml:37]: I did not set those because they are optional, but if we want add all the variables that the role accepts I can do that.

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
+# cloud.aws_ops.move_vm_from_on_prem_to_aws playbooks
+
+A playbook to migrate an existing on prem VM running on KVM hypervisor to AWS.
+
+## Requirements
+
+**qemu** and **qemu-img** packages installed.
+
+The ``cloud.aws_ops.clone_one_prem_vm`` requires privilege escalation because the .qcow2 file created by ``virt-clone`` is owned by root and ``qemu-img convert`` requires access to convert it to .raw.
```
Comment (gravesm) [inline playbooks/move_vm_from_on_prem_to_aws/README.md:9]:
```suggestion
The ``cloud.aws_ops.clone_on_prem_vm`` requires privilege escalation because the .qcow2 file created by ``virt-clone`` is owned by root and ``qemu-img convert`` requires access to convert it to .raw.
```

**Review by gravesm — COMMENTED** [review]
One small thing, otherwise looks good.

File: roles/clone_on_prem_vm/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            ansible_user: "{{ kvm_host.ansible_user }}"
+            ansible_ssh_common_args: -o "UserKnownHostsFile=/dev/null" -o StrictHostKeyChecking=no -i {{ kvm_host.ansible_ssh_private_key_file }}
+            groups: "libvirt"
+
+        - name: Import 'cloud.aws_ops.clone_on_prem_vm' role
+          ansible.builtin.import_role:
+            name: cloud.aws_ops.clone_on_prem_vm
+          vars:
+            clone_on_prem_vm_source_vm_name: "{{ on_prem_source_vm_name }}"
+            clone_on_prem_vm_dest_image_name: "{{ on_prem_vm_image_name }}"
```
Comment (hakbailey) [inline roles/clone_on_prem_vm/README.md:55]: This example uses `clone_on_prem_vm_dest_image_name` but the README says  `clone_on_prem_vm_image_name` and the task looks like it uses `clone_on_prem_vm_image_name` and also `clone_on_prem_vm_dest_vm_name`. These are all supposed to be the same variable right? If so we should make sure the variable is the same in all places in this role and also in the playbook, playbook README, and the `import_image_and_run_aws_instance` role example.

File: roles/clone_on_prem_vm/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    LIBVIRT_DEFAULT_URI: "{{ clone_on_prem_vm_uri }}"
+
+- name: Get information about the clone
+  ansible.builtin.stat:
+    path: "{{ clone_on_prem_vm__clone_path }}"
+  register: clone_on_prem_vm__clone_info
+
+# Priviledge escalation is needed because the .qcow2 file is owned by root
+# when default hypervisor is used
+- name: Convert qcow2 to raw using qemu-img with priviledge escalation
```
Comment (hakbailey) [inline roles/clone_on_prem_vm/tasks/main.yml:92]:
```suggestion
- name: Convert qcow2 to raw using qemu-img with privilege escalation
```

File: roles/import_image_and_run_aws_instance/README.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        * **delete_on_termination_** (bool): Indicates whether the EBS volume is deleted on instance termination.
+
+Dependencies
+------------
+
+- role: [aws_setup_credentials](../aws_setup_credentials/README.md)
+
+Example Playbook
+----------------
+This role can be used together with the [cloud.aws_ops.clone_on_prem_vm](../clone_on_prem_vm/README.md) role as shown below. If you wish to use it separately, just ensure you set the **clone_on_prem_vm_local_image_path** parameter.
```
Comment (hakbailey) [inline roles/import_image_and_run_aws_instance/README.md:61]: I'm not quite sure I follow this. If using this role on its own, wouldn't they just need to set the required `import_image_and_run_aws_instance_image_path` variable?

File: roles/import_image_and_run_aws_instance/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+      when: import_image_and_run_aws_instance_instance_name is undefined
+
+    - name: Get information about the S3 bucket
+      community.aws.s3_bucket_info:
+        name: "{{ import_image_and_run_aws_instance_bucket_name }}"
+      register: import_image_and_run_aws_instance__bucket_info
+
+    - name: Fail when S3 bucket name does not exist
+      ansible.builtin.fail:
+        msg: "The on prem VM {{ import_image_and_run_aws_instance_bucket_name }} does not exist."
```
Comment (hakbailey) [inline roles/import_image_and_run_aws_instance/tasks/main.yml:29]:
```suggestion
    - name: Fail when S3 bucket does not exist
      ansible.builtin.fail:
        msg: "The S3 bucket {{ import_image_and_run_aws_instance_bucket_name }} does not exist."
```

File: roles/import_image_and_run_aws_instance/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    - name: Fail when 'import_image_and_run_aws_instance_image_path' is undefined
+      ansible.builtin.fail:
+        msg: The .raw image path must be defined as import_image_and_run_aws_instance_image_path
+      when: import_image_and_run_aws_instance_image_path is undefined
+
+    - name: Fail when 'import_image_and_run_aws_instance_instance_name' is undefined
+      ansible.builtin.fail:
+        msg: The name to assign to the AWS EC2 instance must be defined as import_image_and_run_aws_instance_instance_name
+      when: import_image_and_run_aws_instance_instance_name is undefined
+
```
Comment (hakbailey) [inline roles/import_image_and_run_aws_instance/tasks/main.yml:21]: Should we also check that the required `import_image_and_run_aws_instance_import_image_task_name` variable is defined?

File: roles/import_image_and_run_aws_instance/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+      until: import_image_and_run_aws_instance__import_image_info.import_image[0].status == "completed"
+      delay: 3
+      retries: 30
+
+    - name: Set 'import_image_and_run_aws_instance__ami_id' and 'import_image_and_run_aws_instance__snapshot_id'
+      ansible.builtin.set_fact:
+        import_image_and_run_aws_instance__ami_id: "{{ import_image_and_run_aws_instance__import_image_info.import_image[0].image_id }}"
+        import_image_and_run_aws_instance__snapshot_id: "{{ import_image_and_run_aws_instance__import_image_info.import_image[0].snapshot_details[0].snapshot_id }}"
+
+    - name: Check if an instance with the specified name alrerady exists
```
Comment (hakbailey) [inline roles/import_image_and_run_aws_instance/tasks/main.yml:72]:
```suggestion
    - name: Check if an instance with the specified name already exists
```

File: roles/import_image_and_run_aws_instance/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+      amazon.aws.ec2_instance_info:
+        filters:
+          "tag:Name": "{{ import_image_and_run_aws_instance_instance_name }}"
+          instance-state-name: [ "running", "pending", "shutting-down", "stopping", "stopped"]
+      register: import_image_and_run_aws_instance__ec2_instance_info
+
+    - name: Fail when an instance with the specified name already exists
```
Comment (hakbailey) [inline roles/import_image_and_run_aws_instance/tasks/main.yml:79]: I wonder if we want to run this check and failure earlier in the role? Since it doesn't look like it depends on any of the previous steps and this is really checking whether the user input for the `import_image_and_run_aws_instance_instance_name` variable will be valid, maybe we give the user that info before the role starts doing other things?

**Review by hakbailey — CHANGES_REQUESTED** [review]

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    * **name**: This is a user-defined name for the host you are adding to the inventory.
+    * **ansible_host**: This variable specifies the hostname or IP address of the host you are adding to the inventory.
+    * **ansible_user**: This variable specifies the SSH username that Ansible should use when connecting to the host.
+    * **ansible_ssh_private_key_file** This variable specifies the path to the SSH private key file that Ansible should use for authentication when connecting to the host.
+    * **groups** This variable enabled you to assign the newly added host to one or more groups in the inventory.
+
+### Needed for the cloud.aws_ops.clone_on_prem_vm role
+
+* **clone_on_prem_vm_source_vm_name** (str): (Required) The name of the on-prem VM you want to clone.
+* **clone_on_prem_vm_image_name** (str): (Optional) The name you want to call the cloned image. If not set, the **on_prem_vm_clone_name** will be used with a _-clone_ suffix.
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/README.md]: Yes, fixed that. I did some lots second changes for variables naming and I forgot to reflect.

File: roles/import_image_and_run_aws_instance/README.md (hunk 7 above)
Comment (alinabuzachis) [reply roles/import_image_and_run_aws_instance/README.md]: I removed that phrase because it feels to me that it brings more confusion. It's clear that the roles can be used separately and what needs to be specified.

File: roles/import_image_and_run_aws_instance/tasks/main.yml (hunk 9 above)
Comment (alinabuzachis) [reply roles/import_image_and_run_aws_instance/tasks/main.yml:21]: Done!

File: roles/import_image_and_run_aws_instance/tasks/main.yml (hunk 11 above)
Comment (alinabuzachis) [reply roles/import_image_and_run_aws_instance/tasks/main.yml]: Moved at the beginning!

File: roles/clone_on_prem_vm/tasks/main.yml (hunk 6 above)
Comment (hakbailey) [reply roles/clone_on_prem_vm/tasks/main.yml:92]: I don't think this was updated either.

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+### Needed for the cloud.aws_ops.clone_on_prem_vm role
+
+* **clone_on_prem_vm_source_vm_name** (str): (Required) The name of the on-prem VM you want to clone.
+* **clone_on_prem_vm_image_name** (str): (Optional) The name you want to call the cloned image. If not set, the **clone_on_prem_vm_source_vm_name** will be used with a _-clone_ suffix.
+* **clone_on_prem_vm_uri** (str): (Optional) Libvirt connection uri. Default: "qemu:///system".
+* **clone_on_prem_vm_overwrite** (bool): (Optional) Whether to overwrite or not an already existing on prem VM clone. Default: true.
+* **clone_on_prem_vm_local_image_path** (str): (Optional) The path where you would like to save the image. If the path does not exists on localhost, the role will create it. If this parameter is not set, the role will save the image in a _~/tmp_ folder.
+
+### Needed for the cloud.aws_ops.import_image_and_run_aws_instance role
```
Comment (abikouo) [inline playbooks/move_vm_from_on_prem_to_aws/README.md:53]: same as above

**Review by abikouo — CHANGES_REQUESTED** [review]

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
+# cloud.aws_ops.move_vm_from_on_prem_to_aws playbooks
+
+A playbook to migrate an existing on prem VM running on KVM hypervisor to AWS.
+
+## Requirements
+
+**qemu** and **qemu-img** packages installed.
+
+The ``cloud.gcp_ops.clone_one_prem_vm`` requires privilege escalation because the .qcow2 file created by ``virt-clone`` is owned by root and ``qemu-img convert`` requires access to convert it to .raw.
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/README.md]: Thanks. It should be now!

File: roles/clone_on_prem_vm/tasks/main.yml (hunk 6 above)
Comment (alinabuzachis) [reply roles/clone_on_prem_vm/tasks/main.yml:92]: Updated now!

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+## Playbook Variables
+
+* **kvm_host** (dict): Information about the host running the KVM hypervisr that are dynamically added to the inventory.
+    * **name**: This is a user-defined name for the host you are adding to the inventory.
+    * **ansible_host**: This variable specifies the hostname or IP address of the host you are adding to the inventory.
+    * **ansible_user**: This variable specifies the SSH username that Ansible should use when connecting to the host.
+    * **ansible_ssh_private_key_file** This variable specifies the path to the SSH private key file that Ansible should use for authentication when connecting to the host.
+    * **groups** This variable enabled you to assign the newly added host to one or more groups in the inventory.
+
+### Needed for the cloud.aws_ops.clone_on_prem_vm role
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/README.md]: Done that. Let me know please if that works!

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+* kms:CreateGrant
+* kms:Decrypt
+* kms:DescribeKey
+* kms:Encrypt
+* kms:GenerateDataKey*
+* kms:ReEncrypt*
+
+## Playbook Variables
+
+* **kvm_host** (dict): Information about the host running the KVM hypervisr that are dynamically added to the inventory.
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/README.md]: Done that. Let me know please if that works!

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+A playbook to migrate an existing on prem VM running on KVM hypervisor to AWS.
+
+## Requirements
+
+This playbook uses the ``cloud.aws_ops.clone_on_prem_vm`` role to clone an existing VM on prem using the KVM hypervisor and the ``cloud.aws_ops.import_image_and_run_aws_instance`` role to import a local .raw image into an Amazon machine image (AMI) and run an AWS EC2 instance. For a complete list of requirements, see [clone_on_prem_vm](../clone_on_prem_vm/README.md#Requirements) and [import_image_and_run_aws_instance](../roles/import_image_and_run_aws_instance/REAME.md#Requirements), respectively.
+
+
+## Playbook Variables
+
+For a fullo list of accepted variables see: [clone_on_prem_vm](../clone_on_prem_vm/README.md#Role-Variables) and respectively [import_image_and_run_aws_instance](../roles/import_image_and_run_aws_instance/REAME.md#Role-Variables).
```
Comment (hakbailey) [inline playbooks/move_vm_from_on_prem_to_aws/README.md:12]:
```suggestion
For a full list of accepted variables see: [clone_on_prem_vm](../clone_on_prem_vm/README.md#Role-Variables) and respectively [import_image_and_run_aws_instance](../roles/import_image_and_run_aws_instance/REAME.md#Role-Variables).
```

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+      ansible_ssh_private_key_file: /path/to/private_key
+      groups: mygroup
+```
+
+All the variables defined in section ``Playbook Variables`` can be defined inside the ``vars/main.yml`` file.
+
+Run the playbook:
+
+```shell
+ansible-playbook cloud.gcp_ops.move_vm_from_on_prem_to_aws -e "@credentials.yml - i inventory.yml"
```
Comment (hakbailey) [inline playbooks/move_vm_from_on_prem_to_aws/README.md:42]:
```suggestion
ansible-playbook cloud.aws_ops.move_vm_from_on_prem_to_aws -e "@credentials.yml - i inventory.yml"
```

**Review by hakbailey — CHANGES_REQUESTED** [review]
Just a couple more small typos. otherwise this looks good!

File: playbooks/move_vm_from_on_prem_to_aws/README.md (hunk 17 above)
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/README.md]: fixed, thanks!

File: playbooks/move_vm_from_on_prem_to_aws/README.md (hunk 18 above)
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/README.md]: fixed, thanks!

**Review by hakbailey — APPROVED** [review]

File: roles/clone_on_prem_vm/README.md [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    kvm:
+      ansible_host: myhost
+      ansible_user: myuser
+      ansible_ssh_private_key_file: /path/to/private_key
+      groups: mygroup
+```
+
+All the variables defined in section ``Playbook Variables`` can be defined inside the ``vars.yml`` file.
+
+Create a ``playbook.ym`` file like this:
```
Comment (GomathiselviS) [inline roles/clone_on_prem_vm/README.md:43]:
```suggestion
Create a ``playbook.yml`` file like this:
```

File: playbooks/move_vm_from_on_prem_to_aws/README.md [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+all:
+  hosts:
+    kvm:
+      ansible_host: myhost
+      ansible_user: myuser
+      ansible_ssh_private_key_file: /path/to/private_key
+      groups: mygroup
+```
+
+All the variables defined in section ``Playbook Variables`` can be defined inside the ``vars.yml`` file.
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply playbooks/move_vm_from_on_prem_to_aws/README.md:37]: Here https://github.com/redhat-cop/cloud.aws_ops/pull/95/files#diff-3a382ac08ed4e75c0e95cbbabd59acfc9e5c8d543f4882cb588ee93f0ad0029fR10. To avoid content duplication I referenced the variable roles which are also the ones required by the playbook.

**Review by GomathiselviS — APPROVED** [review]
